rust/datafusion/src/physical_plan/string_expressions.rs

```diff
@@ -66,3 +71,73 @@ pub fn concatenate(args: &[ArrayRef]) -> Result<StringArray> {
     }
     Ok(builder.finish())
 }
+
+/// character_length returns number of characters in the string
+/// character_length('josé') = 4
+pub fn character_length(args: &[ArrayRef]) -> Result<Int32Array> {
+    let num_rows = args[0].len();
+    let string_args =
+        &args[0]
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .ok_or_else(|| {
+                DataFusionError::Internal(
+                    "could not cast input to StringArray".to_string(),
+                )
+            })?;
+
+    let result = (0..num_rows)
+        .map(|i| {
+            if string_args.is_null(i) {
+                // NB: Since we use the same null bitset as the input,
+                // the output for this value will be ignored, but we
+                // need some value in the array we are building.
+                Ok(0)
+            } else {
+                Ok(string_args.value(i).chars().count() as i32)
+            }
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    let data = ArrayData::new(
+        DataType::Int32,
+        num_rows,
+        Some(string_args.null_count()),
+        string_args.data().null_buffer().cloned(),
+        0,
+        vec![Buffer::from(result.to_byte_slice())],
+        vec![],
+    );
+
+    Ok(Int32Array::from(Arc::new(data)))
+}
+
+macro_rules! string_unary_function {
+    ($NAME:ident, $FUNC:ident) => {
+        /// string function that accepts utf8 and returns utf8
+        pub fn $NAME(args: &[ArrayRef]) -> Result<StringArray> {
+            let string_args = &args[0]
+                .as_any()
+                .downcast_ref::<StringArray>()
+                .ok_or_else(|| {
+                    DataFusionError::Internal(
+                        "could not cast input to StringArray".to_string(),
+                    )
+                })?;
+
+            let mut builder = StringBuilder::new(args.len());
+            for index in 0..args[0].len() {
+                if string_args.is_null(index) {
+                    builder.append_null()?;
+                } else {
+                    builder.append_value(&string_args.value(index).$FUNC())?;
+                }
+            }
+            Ok(builder.finish())
```

Comment:
The Arrow crate implements efficient `IntoIter` and `FromIter` that generally make the code simpler to read and more performant because it performs less bound checks. I.e. something like

```rust
string_args.iter().map(|x| x.map(|x| x.$FUNC())).collect()
// (first map is the iterator, second is for the `Option<_>`
```

will probably work.

Reply:
This did work well but I have struggled to make it work with code that supports both `Utf8` and `LargeUtf8` types as the code does now. Maybe you could help here.

Reply:
The trick is to use the generic `GenericStringArray`, whose `StringArray` and `LargeStringArray` are concrete types of. Something like

```rust
fn op<T: StringOffsetSizeTrait>(array: GenericStringArray<T>) -> GenericStringArray<T> {
     let array = array.downcast::<GenericStringArray<T>>().unwrap();

     array.iter().map(|x| x.map(|x| x.$FUNC())).collect()
}
```

The FromIter and ToIterator are implemented for the generic struct and thus the compiler should be able to resolve these for both `T: i32` and `T: i64`.

Reply:
Thanks @jorgecarleitao . Your code makes a lot of sense and the macro is much cleaner however I am stuck at the next bit which is how to pass in `T`. I can do it in `functions::create_physical_expr` like below but this does not feel correct.

```rust
BuiltinScalarFunction::Lower => {
    |args| {
        match args[0].data_type() {
            DataType::Utf8 => {
                Ok(Arc::new(string_expressions::lower::<i32>(args)?))
            }
            DataType::LargeUtf8 => {
                Ok(Arc::new(string_expressions::lower::<i64>(args)?))
            }
            other => Err(DataFusionError::Internal(format!(
                "Unsupported data type {:?} for function lower",
                other,
            ))),
        }
    }
}
```

Reply:
Ok you can disregard this comment as it is exactly how `length` is implemented. Technically we should not expose `length` as a SQL function but we could rename the alias to be `char_length` and `character_length` based on the ANSI SQL spec: https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#char-length-expression